pkg/controller/snapshotter/snapshotter.go

```diff
@@ -249,93 +351,190 @@ func (vs *volumeSnapshotter) deleteSnapshot(spec *v1.PersistentVolumeSpec, sourc
+	// 1. Check the status of VolumeSnapshot object
+	//	a. Condition Ready or Error -> return nil
+	//	b. Condition Pending (Status True or Unknown) -> step 4
+	//	c. Else -> step 2
+	// 2. Retrieve metadata from VolumeSnapshot object if existing;
+	//	otherwise create metadata and update VolumeSnapshot object.
+	//	If metadata pre-existing, find snapshot through volume plugin;
+	//	If metadata newly created, take snapshot through volume plugin.
+	// 3. Create VolumeSnapshotData object with a reference to the snapshot by the plugin.
+	// 	Update VolumeSnapshot object status and also bind VolumeSnapshotData to VolumeSnapshot.
+	//	Set status to Pending.
+	//	TODO(xyang): Set condition status to True or Unknown based on status from
+	//	querying the snapshot through the volume plugin.
+	// 4. Query the volume plugin for the status of the snapshot with snapshot id
+	//	from VolumeSnapshotData object until snapshot is Ready.
+	//	Update VolumeSnapshot and VolumeSnapshotData, setting status to Ready (True).
+	//	Add the snapshot to ActualStateOfWorld.
 	return func() error {
 		glog.Infof("Enter getSnapshotCreateFunc: snapshotName %s snapshot [%#v]", snapshotName, snapshot)
-		if snapshot.Spec.SnapshotDataName != "" {
-			// update snapshot data status and if complete, adds to asw
-			return vs.updateSnapshotDataStatus(snapshotName, snapshot)
-		}
-		pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
-		if err != nil {
-			return err
-		}
-		pvName := pv.Name
-		snapshotDataSource, err := vs.takeSnapshot(pv)
-		if err != nil || snapshotDataSource == nil {
-			return fmt.Errorf("Failed to take snapshot of the volume %s: %q", pvName, err)
-		}
-		// Snapshot has been created, made an object for it
-		readyCondition := crdv1.VolumeSnapshotDataCondition{
-			Type:    crdv1.VolumeSnapshotDataConditionPending,
-			Status:  v1.ConditionTrue,
-			Message: "Snapshot data is being created",
-		}
-		snapName := "k8s-volume-snapshot-" + string(uuid.NewUUID())
+		//if snapshot.Spec.SnapshotDataName != "" {
+		//	// update snapshot data status and if complete, adds to asw
+		//	return vs.updateSnapshotDataStatus(snapshotName, snapshot)
+		//}
+
+		// Check VolumeSnapshot object status
+		//volSnapshotObj, err := vs.getVolumeSnapshot(snapshotName)
+		index := len(snapshot.Status.Conditions) - 1
+		if len(snapshot.Status.Conditions) > 0 &&
+			((snapshot.Status.Conditions[index].Type == crdv1.VolumeSnapshotConditionReady &&
+			snapshot.Status.Conditions[index].Status == v1.ConditionTrue) ||
+			snapshot.Status.Conditions[index].Type == crdv1.VolumeSnapshotConditionError) {
+			glog.Infof("getSnapshotCreateFunc: Snapshot %s is Ready.", snapshotName)
+			return nil
+		} else if len(snapshot.Status.Conditions) > 0 &&
+			snapshot.Status.Conditions[index].Type == crdv1.VolumeSnapshotConditionPending &&
+			(snapshot.Status.Conditions[index].Status == v1.ConditionTrue ||
+			snapshot.Status.Conditions[index].Status == v1.ConditionUnknown) {
```

Comment:
I wonder what does it mean when setting pending with unknown?

Reply:
I was thinking to convert the "creating" status in Cinder to "pending" with "unknown" because we can't really use the snapshot yet.  For "pending" in AWS, it will be "pending" with "ready", and "uploading" in GCE, it will be "pending" with "ready" as well because the snapshot is already cut at that point.